### ISSUE TITLE
Specify the Inputs-Outputs-Registers of the Datapath and the Controller

### ISSUE THREAD

**Issue by farukyld** [farukyld/sort-circuit#4]
depends on #7 #1

**Comment by farukyld** [comment]
Tasks

- [ ] make the blackbox of the circuit

> 
- [ ] the user interface (start, reset, done, error)
- [ ] the memory interface (relates to #7 )

**Comment by farukyld** [comment]
Tasks

- [ ] make the blackbox of the circuit

> 
> - [ ] the user interface (start, reset, done, error)
> - [ ] the memory interface (relates to #7 )

**Comment by farukyld** [comment]
Tasks

- [x] define the external connections of the circuit
> - [x] the clock
> - [x] the user interface (start, reset, done, error)
> - [x] the memory interface (relates to #7 )

- [x] relate external connections to datapath or controller
- [x] define registers (as a template) of the datapath
- [x] define interconnections (as a template) between controller and datapath.

**Comment by farukyld** [comment]
I am going to write [verilog codes](https://github.com/farukyld/sort-circuit/tree/main/verilog_codes) to define those.

**Comment by farukyld** [comment]
I applied this style [guide](https://github.com/lowRISC/style-guides/blob/master/VerilogCodingStyle.md#appendix---condensed-style-guide). and this [document](https://inst.eecs.berkeley.edu/~cs150/Documents/Nets.pdf) helped me remembering the usecases of reg and wire variables.

**Comment by farukyld** [comment]
I first wrote the external pins of the [sort_circuit](https://github.com/farukyld/sort-circuit/blob/4-specify-the-inputs-outputs-registers-of-the-datapath-and-the-controller/verilog_codes/sort_circuit.v). this module acts as a container for [datapath](https://github.com/farukyld/sort-circuit/blob/4-specify-the-inputs-outputs-registers-of-the-datapath-and-the-controller/verilog_codes/datapath.v) and [controller](https://github.com/farukyld/sort-circuit/blob/4-specify-the-inputs-outputs-registers-of-the-datapath-and-the-controller/verilog_codes/sort_circuit.v).

Then I inputs/outputs of datapath. after that, I asked chatGPT to give me the controller's in/out pins. I know that AI is good at those kind of automated jobs. It is actuallybetter than humans in such repetitive labor. humans (at least me) tend to make more mistakes in jobs relying on copy-paste.

It created the header of the module controller, and I asked it to make neccessary wiring and module instantiation. Again, AI is very talented at this job. Then I made some little corrections.
